subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncher.java

```diff
@@ -201,6 +147,75 @@ public void stop() {
         }
     }
 
+    private class RunBuildAction implements Transformer<BuildResult, BuildOperationContext> {
+        private final Stage upTo;
+
+        private RunBuildAction(Stage upTo) {
+            this.upTo = upTo;
+        }
+
+        @Override
+        public BuildResult transform(BuildOperationContext buildOperationContext) {
+            BuildOperationWorkerRegistry.Completion workerLease = buildServices.get(BuildOperationWorkerRegistry.class).operationStart();
```

Comment:
I think this API would benefit from a SAM type action:

`workerRegistry.operation { ... }`

this would prevent from accidental forget of `operationFinish`.